File: TMessagesProj/src/main/java/org/telegram/messenger/fakepasscode/AccountActions.java
package org.telegram.messenger.fakepasscode;

import org.telegram.messenger.SharedConfig;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class AccountActions {
    public int accountNum;
    private final FakePasscode fakePasscode;

    private <T extends AccountAction> T getAction(List<T> actions) {
        return actions.stream().filter(a -> a.accountNum == accountNum).findFirst().orElse(null);
    }

    private <T extends AccountAction> T getOrCreateAction(List<T> actions, Class<T> clazz) {
        return actions.stream().filter(a -> a.accountNum == accountNum)
                .findFirst().orElseGet(() -> {
                    try {
                        T action = clazz.newInstance();
                        action.accountNum = accountNum;
                        actions.add(action);
                        return action;
                    } catch (Exception ignored) {
                        return null;
                    }
                });
    }

    private <T extends AccountAction> void toggleAction(List<T> actions, Class<T> clazz) {
        T action = getAction(actions);
        if (action != null) {
            actions.removeIf(a -> a.accountNum == accountNum);
        } else {
            try {
                action = clazz.newInstance();
                action.accountNum = accountNum;
                actions.add(action);
            } catch (Exception ignored) {
            }
        }
        SharedConfig.saveConfig();
    }

    public RemoveChatsAction getRemoveChatsAction() { return getOrCreateAction(fakePasscode.removeChatsActions, RemoveChatsAction.class); }
    public TelegramMessageAction getMessageAction() { return getOrCreateAction(fakePasscode.telegramMessageAction, TelegramMessageAction.class); }

    public ArrayList<Integer> getChatsToRemove() {
        return getRemoveChatsAction().chatsToRemove;
    }

    public void setChatsToRemove(ArrayList<Integer> chats) {
        getRemoveChatsAction().chatsToRemove = chats;
        SharedConfig.saveConfig();
    }

    public void toggleDeleteContactsAction() { toggleAction(fakePasscode.deleteContactsActions, DeleteContactsAction.class); }
    public void toggleDeleteStickersAction() { toggleAction(fakePasscode.deleteStickersActions, DeleteStickersAction.class); }
    public void toggleClearSearchHistoryAction() { toggleAction(fakePasscode.clearSearchHistoryActions, ClearSearchHistoryAction.class); }
    public void toggleTerminateOtherSessionsAction() { toggleAction(fakePasscode.terminateOtherSessionsActions, TerminateOtherSessionsAction.class); }
    public void toggleLogOutAction() { toggleAction(fakePasscode.logOutActions, LogOutAction.class); }

    public boolean isDeleteContacts() { return getAction(fakePasscode.deleteContactsActions) != null; }
    public boolean isDeleteStickers() { return getAction(fakePasscode.deleteStickersActions) != null; }
    public boolean isClearSearchHistory() { return getAction(fakePasscode.clearSearchHistoryActions) != null; }
    public boolean isTerminateOtherSessions() { return getAction(fakePasscode.terminateOtherSessionsActions) != null; }
    public boolean isLogOut() { return getAction(fakePasscode.logOutActions) != null; }

    public AccountActions(int accountNum, FakePasscode fakePasscode)
    {
        this.accountNum = accountNum;
        this.fakePasscode = fakePasscode;
    }

    public String getPhone() {
        return fakePasscode.phoneNumbers.getOrDefault(accountNum, "");
    }

    public void setPhone(String phone) {
        fakePasscode.phoneNumbers.put(accountNum, phone);
    }
}

File: TMessagesProj/src/main/java/org/telegram/messenger/fakepasscode/AccountActions.kt
package org.telegram.messenger.fakepasscode

import android.util.Base64
import org.telegram.messenger.AccountInstance
import org.telegram.messenger.SharedConfig
import org.telegram.messenger.UserConfig
import org.telegram.messenger.Utilities
import org.telegram.tgnet.TLRPC
import kotlin.reflect.KMutableProperty
import kotlin.reflect.full.memberProperties
import kotlin.reflect.jvm.isAccessible
import kotlin.reflect.jvm.javaType

class AccountActions : Action {
    var accountNum: Int? = null
    var removeChatsAction = RemoveChatsAction()
        private set(value) { field = value; SharedConfig.saveConfig() }
    var telegramMessageAction = TelegramMessageAction()
        private set(value) { field = value; SharedConfig.saveConfig() }
    private var deleteContactsAction: DeleteContactsAction? = null
        private set(value) { field = value; SharedConfig.saveConfig() }
    private var deleteStickersAction: DeleteStickersAction? = null
        private set(value) { field = value; SharedConfig.saveConfig() }
    private var clearSearchHistoryAction: ClearSearchHistoryAction? = null
        private set(value) { field = value; SharedConfig.saveConfig() }
    private var clearBlackListAction: ClearBlackListAction? = null
        private set(value) { field = value; SharedConfig.saveConfig() }
    private var clearSavedChannelsAction: ClearSavedChannelsAction? = null
        private set(value) { field = value; SharedConfig.saveConfig() }
    private var clearDraftsAction: ClearDraftsAction? = null
        private set(value) { field = value; SharedConfig.saveConfig() }
    var terminateOtherSessionsAction = TerminateOtherSessionsAction()
        private set(value) { field = value; SharedConfig.saveConfig() }
    private var logOutAction: LogOutAction? = null
        private set(value) { field = value; SharedConfig.saveConfig() }
    private var hideAccountAction: HideAccountAction? = null
        private set(value) { field = value; SharedConfig.saveConfig() }
    var fakePhone = ""
        set(value) { field = value; SharedConfig.saveConfig() }
    var sessionsToHide = CheckedSessions()
        private set(value) { field = value; SharedConfig.saveConfig() }
    private var salt: String? = null
        private set(value) { field = value; SharedConfig.saveConfig() }
    var idHash: String? = null
        private set(value) { field = value; SharedConfig.saveConfig() }

    companion object {
        var updateIdHashEnabled = false
    }

    init {
        if (updateIdHashEnabled) {
            Utilities.globalQueue.postRunnable(UpdateIdHashRunnable(this), 1000)
        }
    }

    override fun execute(fakePasscode: FakePasscode) {
        accountNum?.let { acc ->
            listOfNotNull(
                removeChatsAction, telegramMessageAction, deleteContactsAction,
                deleteStickersAction, clearSearchHistoryAction, clearBlackListAction,
                clearSavedChannelsAction, clearDraftsAction, terminateOtherSessionsAction,
                logOutAction, hideAccountAction
            ).forEach { action ->
                    action.setAccountNum(acc)
                    action.execute(fakePasscode)
            }
        }
    }

    private fun getSalt(): ByteArray {
        if (salt == null) {
            val saltBytes = ByteArray(16)
            Utilities.random.nextBytes(saltBytes)
            salt = Base64.encodeToString(saltBytes, Base64.DEFAULT)
        }
        return Base64.decode(salt, Base64.DEFAULT)
    }

    private fun calculateIdHash(user: TLRPC.User): String {
        val phoneDigits = user.phone.filter { it.isDigit() }
        if (phoneDigits.length < 4) {
            throw Exception("Can't calculate id hash: invalid phone")
        }
        val phoneId = phoneDigits.slice(phoneDigits.length - 4 until phoneDigits.length).toInt()
        val sum = (user.id % 10_000 + phoneId) % 10_000
        val sumBytes = sum.toString().encodeToByteArray()
        val bytes = ByteArray(32 + sumBytes.size)
        val salt = getSalt()
        System.arraycopy(salt, 0, bytes, 0, 16)
        System.arraycopy(sumBytes, 0, bytes, 16, sumBytes.size)
        System.arraycopy(salt, 0, bytes, sumBytes.size + 16, 16)
        return Utilities.bytesToHex(Utilities.computeSHA256(bytes, 0, bytes.size))
    }

    fun checkIdHash() {
        if (accountNum != null) {
            accountNum?.let { account ->
                val userConfig = UserConfig.getInstance(account)
                if (userConfig.isClientActivated) {
                    idHash = calculateIdHash(userConfig.currentUser)
                } else {
                    accountNum = null
                }
            }
        } else {
            checkAccountNum(false)
        }
    }

    fun checkAccountNum(allowSwapAccountNum: Boolean) {
        if (idHash != null) {
            for (a in 0 until UserConfig.MAX_ACCOUNT_COUNT) {
                val userConfig = UserConfig.getInstance(a)
                if (userConfig.isClientActivated) {
                    if (idHash == calculateIdHash(userConfig.currentUser)) {
                        for (fakePasscode in SharedConfig.fakePasscodes) {
                            val actions = fakePasscode.accountActions
                            if (actions.contains(this) && (allowSwapAccountNum
                                        || actions.stream().noneMatch { it.accountNum == a })) {
                                accountNum = a
                                break
                            }
                        }
                        break
                    }
                }
            }
        }
    }

    inline fun <reified T> reverse(action: T?): T? {
        if (action != null) {
            return null
        }
        return T::class.constructors.find { it.parameters.isEmpty() }?.call()
    }

    fun toggleDeleteContactsAction() { deleteContactsAction = reverse(deleteContactsAction) }
    fun toggleDeleteStickersAction() { deleteStickersAction = reverse(deleteStickersAction) }
    fun toggleClearSearchHistoryAction() { clearSearchHistoryAction = reverse(clearSearchHistoryAction) }
    fun toggleClearBlackListAction() { clearBlackListAction = reverse(clearBlackListAction) }
    fun toggleClearSavedChannelsAction() { clearSavedChannelsAction = reverse(clearSavedChannelsAction) }
    fun toggleClearDraftsAction() { clearDraftsAction = reverse(clearDraftsAction) }
    fun toggleLogOutAction() { logOutAction = reverse(logOutAction) }
    fun toggleHideAccountAction() { hideAccountAction = reverse(hideAccountAction) }

    fun setSessionsToHide(sessions: List<Long?>?) {
        sessionsToHide.sessions = sessions
        SharedConfig.saveConfig()
    }

    fun isDeleteContacts() = deleteContactsAction != null
    fun isDeleteStickers() = deleteStickersAction != null
    fun isClearSearchHistory() = clearSearchHistoryAction != null
    fun isClearBlackList() = clearBlackListAction != null
    fun isClearSavedChannels() = clearSavedChannelsAction != null
    fun isClearDraftsAction() = clearDraftsAction != null
    fun isLogOut() = logOutAction != null
    fun isHideAccount() = hideAccountAction != null
    fun isLogOutOrHideAccount() = logOutAction != null || hideAccountAction != null

    fun getChatsToRemoveCount() = removeChatsAction.chatEntriesToRemove.size

    fun removeFakePhone() {
        fakePhone = ""
    }

    fun <T: AccountAction> setAction(action: T) {
        for (property in AccountActions::class.memberProperties) {
            if (property.returnType.javaType == action.javaClass && property is KMutableProperty<*>) {
                property.isAccessible = true
                property.setter.call(this, action);
            }
        }
    }
}